pkg/dns/dnsprovider/dnsProvider.go

```
@@ -0,0 +1,104 @@
+	dnsprovider := &v1alpha1.DNSProvider{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      managedZone.Spec.ProviderRef.Name,
+			Namespace: managedZone.Spec.ProviderRef.Namespace,
+		}}
+	log.Log.Info("Reconciling DNS Provider:", "Name:", dnsprovider.Name)
+	err := p.client.Get(ctx, client.ObjectKeyFromObject(dnsprovider), dnsprovider)
+	if err != nil {
+		return nil, err
+	}
+
+	providerSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dnsprovider.Spec.Credentials.Name,
+			Namespace: dnsprovider.Spec.Credentials.Namespace,
+		}}
+	log.Log.Info("Reconciling DNS provider secret:", "secretName", providerSecret.Name)
+	err = p.client.Get(ctx, client.ObjectKeyFromObject(providerSecret), providerSecret)
+	if err != nil {
+		return nil, err
+	}
+	aws := &v1alpha1.DNSProviderConfig{}
+
+	if len(providerSecret.Data) == 0 {
+		log.Log.Info("Secret is empty")
+		return nil, nil
+	}
+	if aws.Route53 == nil {
+		aws.Route53 = &v1alpha1.DNSProviderConfigRoute53{}
+	}
+
+	if managedZone.Spec.ProviderRef.ProviderType == "AWS" {
+
+		if value, ok := providerSecret.Data["AWS_SECRET_ACCESS_KEY"]; ok {
+			aws.Route53.SecretAccessKey = string(value)
+		}
+
+		if value, ok := providerSecret.Data["AWS_ACCESS_KEY_ID"]; ok {
+			aws.Route53.AccessKeyID = string(value)
+		}
+
+		if value, ok := providerSecret.Data["REGION"]; ok {
+			aws.Route53.Region = string(value)
+		}
+	}
+
+	return aws, nil
+}
+
+func (p *provider) CreateDNSProvider(ctx context.Context, managedZone *v1alpha1.ManagedZone) (dns.Provider, error) {
+	creds, err := p.dnsProviderSecret(ctx, managedZone)
+	if err != nil {
+		return nil, err
+	}
+
+	switch managedZone.Spec.ProviderRef.ProviderType {
+	case "AWS":
+		log.Log.Info("Creating DNS provider for provider type AWS")
+		DNSProvider, _ = aws.NewDNSProvider(*creds)
+		return DNSProvider, nil
+
+	case "GCP":
+		log.Log.Info("GCP")
+
+	case "AZURE":
+		log.Log.Info("AZURE")
+
+	default:
```

I think here I would return an error UnsupportedDNSProvider